roachtest: multitenant-upgrade failed

roachtest.multitenant-upgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2945603&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2945603&tab=artifacts#/multitenant-upgrade) on master @ [dbf711a83969210bcf706d65f53f888856da6ad0](https://github.com/cockroachdb/cockroach/commits/dbf711a83969210bcf706d65f53f888856da6ad0):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/multitenant-upgrade/run_1
	test.go:298,assertions.go:262,assertions.go:1332,require.go:1035,multitenant_upgrade.go:136,multitenant_upgrade.go:35,test_runner.go:777: 
			Error Trace:	multitenant_upgrade.go:136
			            				multitenant_upgrade.go:35
			            				test_runner.go:777
			            				asm_amd64.s:1374
			Error:      	Received unexpected error:
			            	prev version not set for version: 21.2
			            	(1) attached stack trace
			            	  -- stack trace:
			            	  | main.PredecessorVersion
			            	  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:1210
			            	  | main.runMultiTenantUpgrade
			            	  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/multitenant_upgrade.go:135
			            	  | main.registerMultiTenantUpgrade.func1
			            	  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/multitenant_upgrade.go:35
			            	  | main.(*testRunner).runTest.func2
			            	  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:777
			            	  | runtime.goexit
			            	  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
			            	Wraps: (2) prev version not set for version: 21.2
			            	Error types: (1) *withstack.withStack (2) *errutil.leafError
			Test:       	multitenant-upgrade
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh multitenant-upgrade
```
</p>
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*multitenant-upgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.multitenant-upgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2949834&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2949834&tab=artifacts#/multitenant-upgrade) on master @ [f9dca40d3afb0a605eeee76bec4a07ff65d6fec0](https://github.com/cockroachdb/cockroach/commits/f9dca40d3afb0a605eeee76bec4a07ff65d6fec0):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/multitenant-upgrade/run_1
	sql_runner.go:59,multitenant_upgrade.go:168,multitenant_upgrade.go:35,test_runner.go:777: error executing 'SET CLUSTER SETTING cluster.preserve_downgrade_option = '20.2'': pq: cannot set cluster.preserve_downgrade_option to 20.2 (cluster version is 20.2-48)

	cluster.go:2340,multitenant_upgrade.go:80,panic.go:969,test.go:308,test.go:288,sql_runner.go:59,multitenant_upgrade.go:168,multitenant_upgrade.go:35,test_runner.go:777: output in run_231747.781_n2_pkill: cluster.RunL: context canceled
		(1) attached stack trace
		  -- stack trace:
		  | main.(*cluster).RunE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2418
		  | [...repeated from below...]
		Wraps: (2) output in run_231747.781_n2_pkill
		Wraps: (3) attached stack trace
		  -- stack trace:
		  | main.(*cluster).RunL
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2424
		  | main.(*cluster).RunE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2405
		  | main.(*cluster).Run
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2338
		  | main.(*tenantNode).stop
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/multitenant_upgrade.go:80
		  | runtime.gopanic
		  | 	/usr/local/go/src/runtime/panic.go:969
		  | main.(*test).fatalfInner
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test.go:308
		  | main.(*test).Fatalf
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test.go:288
		  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).Exec
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:59
		  | main.runMultiTenantUpgrade
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/multitenant_upgrade.go:168
		  | main.registerMultiTenantUpgrade.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/multitenant_upgrade.go:35
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:777
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (4) cluster.RunL
		Wraps: (5) context canceled
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *errors.errorString
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh multitenant-upgrade
```
</p>
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*multitenant-upgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cmd/roachtest/multitenant_upgrade.go
+++ b/pkg/cmd/roachtest/multitenant_upgrade.go
@@ -165,7 +165,13 @@ func runMultiTenantUpgrade(ctx context.Context, t *test, c *cluster, v version.V
 			withResults([][]string{{"1", "bar"}}))
 
 	t.Status("preserving downgrade option on host server")
-	runner.Exec(t, `SET CLUSTER SETTING cluster.preserve_downgrade_option = '20.2'`)
+	{
+		s := runner.QueryStr(t, `SHOW CLUSTER SETTING version`)
+		runner.Exec(
+			t,
+			`SET CLUSTER SETTING cluster.preserve_downgrade_option = $1`, s[0][0],
+		)
+	}
 
 	t.Status("upgrading host server")
 	c.Stop(ctx, kvNodes)
@@ -230,12 +236,29 @@ func runMultiTenantUpgrade(ctx context.Context, t *test, c *cluster, v version.V
 	t.Status("starting the tenant 11 server with the current binary")
 	tenant11.start(ctx, t, c, currentBinary)
 
+	if v, err := version.Parse("v" + predecessor); err != nil {
+		t.Fatal(err)
+	} else if !v.AtLeast(version.MustParse("v21.1.1")) {
+		// If we take this branch, the current branch is 21.2 but we haven't released 21.1.1 yet.
+		// The multi-tenant upgrade PR #60730 will only be contained in 21.1.1. That PR is responsible
+		// for giving tenants the initial cluster version, which is thus absent here. We still want to
+		// run this test until #60730 is backported, so fill the cluster version in. Without this, the
+		// tenant would report a cluster version of 20.2 which is going to trip up the test. We won't
+		// actually upgrade tenants in production in scenarios in which this branch is active (the host
+		// cluster will be at the latest patch release before attempting an upgrade, which is going to
+		// be at least 21.1.1).
+		// This can be deleted once the PredecessorVersion(21.2) is >= 21.1.1.
+		verifySQL(t, tenant11.pgURL, mkStmt(`SET CLUSTER SETTING version = $1`, initialVersion))
+	}
+
 	t.Status("verify tenant 11 server works with the new binary")
-	verifySQL(t, tenant11.pgURL,
-		mkStmt(`SELECT * FROM foo LIMIT 1`).
-			withResults([][]string{{"1", "bar"}}),
-		mkStmt("SHOW CLUSTER SETTING version").
-			withResults([][]string{{initialVersion}}))
+	{
+		verifySQL(t, tenant11.pgURL,
+			mkStmt(`SELECT * FROM foo LIMIT 1`).
+				withResults([][]string{{"1", "bar"}}),
+			mkStmt("SHOW CLUSTER SETTING version").
+				withResults([][]string{{initialVersion}}))
+	}
 
 	// Note that this is exercising a path we likely want to eliminate in the
 	// future where the tenant is upgraded before the KV nodes.

